email/exporter_test.go

```diff
@@ -0,0 +1,122 @@
+package email
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	emailpb "github.com/letsencrypt/boulder/email/proto"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/metrics"
+	"github.com/letsencrypt/boulder/test"
+)
+
+var ctx = context.Background()
+
+// MockPardotClientImpl is a mock implementation of PardotClient.
+type MockPardotClientImpl struct {
+	sync.Mutex
+	CreatedContacts []string
+}
+
+// NewMockPardotClientImpl returns a MockPardotClientImpl, implementing the
+// PardotClient interface. Both refer to the same instance, with the interface
+// for mock interaction and the struct for state inspection and modification.
+func NewMockPardotClientImpl() (PardotClient, *MockPardotClientImpl) {
+	mockImpl := &MockPardotClientImpl{
+		CreatedContacts: []string{},
+	}
+	return mockImpl, mockImpl
+}
+
+// SendContact adds an email to CreatedContacts.
+func (m *MockPardotClientImpl) SendContact(email string) error {
+	m.Lock()
+	defer m.Unlock()
+
+	m.CreatedContacts = append(m.CreatedContacts, email)
+	return nil
+}
+
+func (m *MockPardotClientImpl) getCreatedContacts() []string {
+	m.Lock()
+	defer m.Unlock()
+	// Return a copy to avoid race conditions.
+	return append([]string(nil), m.CreatedContacts...)
+}
+
+// setup creates a new ExporterImpl, a MockPardotClientImpl, and the start and
+// cleanup functions for the ExporterImpl. Call start() to begin processing the
+// ExporterImpl queue and cleanup() to drain and shutdown. If start() is called,
+// cleanup() must be called.
+func setup() (*ExporterImpl, *MockPardotClientImpl, func(), func()) {
+	mockClient, clientImpl := NewMockPardotClientImpl()
+	exporter := NewExporterImpl(mockClient, 1000000, 5, metrics.NoopRegisterer, blog.NewMock())
+	daemonCtx, cancel := context.WithCancel(context.Background())
+	return exporter, clientImpl,
+		func() { exporter.Start(daemonCtx) },
+		func() {
+			cancel()
+			exporter.Drain()
+		}
+}
+
+func TestSendContacts(t *testing.T) {
+	t.Parallel()
+
+	exporter, clientImpl, start, cleanup := setup()
+	start()
+	defer cleanup()
+
+	_, err := exporter.SendContacts(ctx, &emailpb.SendContactsRequest{
+		Emails: []string{"test@example.com", "user@example.com"},
+	})
+
+	// Wait for the queue to be processed.
+	time.Sleep(100 * time.Millisecond)
+
+	test.AssertNotError(t, err, "Error creating contacts")
+	test.AssertEquals(t, 2, len(clientImpl.getCreatedContacts()))
```

Comment:
Similar to the feedback on the other PR, we should avoid sleeps in our tests whenever possible.

We can assert on the `err` immediately since that's set immediately.

For the `len(clientImpl.getCreatedContacts())`, we can loop until it reaches the len we want. Or, better: we can loop until it contains both emails that we want it to contain.